### DISCUSSION TITLE
How to inject Data to an navigated Page-Viewmodel using NavigationView and Regions

### DISCUSSION THREAD

**Discussion by VRage** [unoplatform/uno.extensions#2413] [Q&A] [answered]
Hello, 
for simplicity i'll use [this](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/Advanced/HowTo-UseNavigationView.html#navigating-to-page-elements) article to explain my problem. I try to "transfere" data from `MainViewModel` to `ProjectsViewModel`. Therefore i used `uen:Navigation.Data` in `MainPage.xaml`:
```xaml
  <Grid uen:Region.Attached="True">
    <NavigationView uen:Region.Attached="true">
     <NavigationView.MenuItems>
        <NavigationViewItem Content="Products"
                                    uen:Region.Name="Project"
                                    uen:Navigation.Data="{Binding Name, Mode=TwoWay}"/>
      </NavigationView.MenuItems>
``` 
The `Name` property is defined in MainViewModel:
```c#
    [ObservableProperty]
    private string? name;
``` 
For the Projectspage i've defined a `DataViewMap` in `RegisterRoutes` method:
```c#
    private static void RegisterRoutes(IViewRegistry views, IRouteRegistry routes)
    {
        views.Register(
            new ViewMap(ViewModel: typeof(ShellViewModel)),
            new ViewMap<MainPage, MainViewModel>(),
            new DataViewMap<ProjectPage, ProjectViewModel, string>()
        );

        routes.Register(
            new RouteMap("", View: views.FindByViewModel<ShellViewModel>(),
                Nested:
                [
                    new ("Main", View: views.FindByViewModel<MainViewModel>()),
                    new ("Project", View: views.FindByViewModel<ProjectViewModel>()),
                ]
            )
        );
    }
``` 
I get a Breakpoint on the contructor for ProjectViewModel but the string is null.
In the [ToDo app](https://github.com/unoplatform/Uno.Samples/blob/5e27ec67ca0ec615d3c8edc95bb7194f372b5a8f/reference/ToDo/src/ToDo/Views/HomePage.xaml#L114) there is something similar, but i can't get it to work on my code.

Of Course this is a Simple example. My goal is to initialize Objects in MainViewModel and to inject these Data to VMs of other pages i want to navigate through NavigationView.

**Comment by VRage** [comment, marked as answer]
Seems like if `uen:Region.Name="Project"` is replaced with `uen:Navigation.Request="./Project"` it kinda works. Still don't know why  `uen:Region.Name` ignores the `Navigation.Data`

**Comment by francoistanguay** [comment]
Have you confirmed that it works if you use ViewModel-based navigation?

Meaning you could define a Command in your MainModel that is bound using CommandExtensions.Command on your NavigationView: https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/helpers/command-extensions.html#breadcrumb

And then in your command you Navigate using NavigateViewModelAsync/DataAsync: https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-NavigateInCode.html

?

**Reply by VRage** [reply]
No i didn't, but this will work for sure. Code-behind navigation also works. But Region based navigation is a neat solution to navigate through pages, just would be great to attach data.

**Reply by francoistanguay** [reply]
You can use region-based navigation with .NavigateRouteAsync and would think you can pass-in data/viewmodel. We probably need better documentation on this.

What is not currently supported is to provide the data using Navigation.Data when not using Navigation.Request to trigger the navigation. They go hand-in-hand. So Navigation.Request+Data mostly used when wanting to do XAML-based navigation from a Button to a new destination.

**Reply by VRage** [reply]
Ok, but now i know. I thinks it's not the fault auf documentation, i've just expected a functionality. I Just though, that there must been a way the transfere data. One more thing tho. I've defined a Page in the RoutMap as default, but I'ld to transfere objects to this viewmodel from MainViewModel, the NavigationData is null. But I guess i have to use  ` .NavigateRouteAsync` again. It's kinda "weird" that the Navigation.Data is ignored in some cases.

**Reply by francoistanguay** [reply]
Yeah, I get the expectation. ALl of this started with Navigation.Request being added for xaml-based navigation and quickly realized that we needed to pass data and there came Navigation.Data.

In your specific scenario, one could argue that NavigationView is a "root" level kind of control and it's mostly used to do region-based navigation to root pages, without the need to pass data. If there's ambient data required, it's typically fetched through a "singleton view model" registered in the DI, or through a Service layer.

Again, your scenario is valid, we just never thought of it or prioritized it so that navigation that can be triggered by controls themselves, thanks to Region.Attached only on NavigationView, could still bind/pass local data using Navigation.Data.

I would assume it shouldn't be too complicated to update the support of the Navigation Adapter for NavigationView that gets spawned when you set Region.Attached to also support Navigation.Data. Who knows, you might be able to achieve it yourself before the team gets to attend to it :)
